### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by sbsce** [rhasspy/piper#24]
Would be nice if this would officially support Windows. So far, there's only a release for "desktop Linux" and "Raspberry Pi 4", not for Windows.

**Comment by synesthesiam** [comment]
I've started working on a Rust version that I hope will make supporting Windows easier. If you have any C++ experience on Windows, any tips would be appreciated.

**Comment by sbsce** [comment]
Well I need this as a library in my C++ program, so I certainly like this being C++ more than Rust ;)

Windows support should be easy, should be just a matter of using cmake to generate Visual Studio project files.

**Comment by synesthesiam** [comment]
The only two dependencies are espeak-ng and the onnxruntime, so the Visual Studio Project could probably be created by hand pretty easily. It's been so many years that I wouldn't know where to start anymore :smile:

**Comment by Wetzel402** [comment]
You could try WSL or Docker.

**Comment by synesthesiam** [comment]
Windows support is still a work in progress

**Comment by coffeecodeconverter** [comment]
so ive got this working in WSL for windows. 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Windows\System32>wsl --version
WSL version: 2.3.26.0
Kernel version: 5.15.167.4-1
WSLg version: 1.0.65
MSRDC version: 1.2.5620
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26100.1-240331-1435.ge-release
Windows version: 10.0.22621.3374
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

i have a RTX 4060 laptop GPU 8GB VRAM
64GB RAM
Ryzen 7 7840hs 

CPU will train 1 x epoch in approximately 51 secs  
GPU will train 1 x epoch in approximately 26 secs

im still tweaking some things to streamline an install on windows, as there were many moving parts so to speak, 
but ive got it fully working, so it can be done, and when it works, its pretty sweet. 
so i'll re-post a link shortly
as i've included troubleshooting all the various warning messages as well, which most tutorials skip over, 
and ive tweaked some of the pytorch lightning .py files to include timestamps during training to make the output more readable. 
~

**Comment by coffeecodeconverter** [comment]
**INSTALL WSL**
1. Open CMD, as administrator, and type:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wsl --install Ubuntu-22.04
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

2. Create a User and Password 

3. Test WSL by running this command:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wsl --version
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

should return a WSL version thats 2.0 or higher (WSL 1.0 wont work for this) 
Example output:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
WSL version: 2.3.26.0        <<<<< Higher than 2.0, we're good
Kernel version: 5.15.167.4-1
WSLg version: 1.0.65
MSRDC version: 1.2.5620
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26100.1-240331-1435.ge-release
Windows version: 10.0.22621.3374
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


4. run this command to update Ubuntu
~~~~~~~~~~~~~~~~~~~~~~~~~~
sudo apt update 
~~~~~~~~~~~~~~~~~~~~~~~~~~
.
.
.
.






**INSTALL PYTHON 3.10**

1. run this command to install python 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sudo spt install python3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

2. run this command 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sudo apt install python3.10-venv
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.
.
.
.



**INSTALL CUDA TOOLKIT**
1. install the Cuda Tool Kit within the WSL 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sudo apt install cuda-toolkit-11-8
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.
.
.
.





**INSTALL PIPER TTS & PIPER-TRAIN**
1. firstly, make sure you're at your root directory with this command
~~~~~~
cd ~
~~~~~~

2. create a new directory, such as "training" 
~~~~~~~~~~~~~~~~~~~~~~~~~
mkdir training
~~~~~~~~~~~~~~~~~~~~~~~~~

3. navigate into the new directory 
~~~~~~~~~~~~~~~~~~~~~~~~~
cd training
~~~~~~~~~~~~~~~~~~~~~~~~~

4. download piper tts (which includes the piper-train as well) 
~~~~~~~~~~~~~~~~~~~~~~~~~
git clone https://github.comrhasspy/piper.git
~~~~~~~~~~~~~~~~~~~~~~~~~

5. create a virtual environment 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
python3 -m venv .venv
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

6. activate the virtual environment 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
source .venv/bin/activate
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

7. install specific version of pip 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
python3 -m pip install pip==23.3.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

8. install specific version of numpy 1.24.4
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
python3 -m pip install numpy==1.24.4
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

9. navigate into the piper src folder
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cd ~/traning/piper/src/python 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

10. backup the existing requirements, and replace them with these:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pip==23.3.1
numpy==1.24.4
torchmetrics==0.11.4
wheel==0.45.1
setuptools==75.6.0
cython>=0.29.0,<1
librosa>=0.9.2,<1
piper-phonemize==1.1.0
onnxruntime>=1.11.0
pytorch-lightning==1.9.0
onnx
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

10. install the requirements 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pip3 install -e .
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

11. install specific torch cuda version 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pip install torch==2.0.0+cu118 torchaudio==2.0.0 -f https://download.pytorch.org/whl/torch_stable.html
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

12. hopefully, your pip list should look like this:
~~~~~~~~~~~~~~~~~~~~~~~~~~~
absl-py                  2.1.0
aiohappyeyeballs         2.4.3
aiohttp                  3.11.7
aiosignal                1.3.1
async-timeout            5.0.1
attrs                    24.2.0
audioread                3.0.1
certifi                  2024.8.30
cffi                     1.17.1
charset-normalizer       3.4.0
cmake                    3.31.1
coloredlogs              15.0.1
Cython                   0.29.37
decorator                5.1.1
filelock                 3.16.1
flatbuffers              24.3.25
frozenlist               1.5.0
fsspec                   2024.10.0
grpcio                   1.68.0
humanfriendly            10.0
idna                     3.10
Jinja2                   3.1.4
joblib                   1.4.2
lazy_loader              0.4
librosa                  0.10.2.post1
lightning-utilities      0.11.9
lit                      18.1.8
llvmlite                 0.43.0
Markdown                 3.7
MarkupSafe               3.0.2
mpmath                   1.3.0
msgpack                  1.1.0
multidict                6.1.0
networkx                 3.4.2
numba                    0.60.0
numpy                    1.24.4
nvidia-cublas-cu11       11.11.3.6
nvidia-cublas-cu12       12.4.5.8
nvidia-cuda-cupti-cu11   11.8.87
nvidia-cuda-cupti-cu12   12.4.127
nvidia-cuda-nvrtc-cu11   11.8.89
nvidia-cuda-nvrtc-cu12   12.4.127
nvidia-cuda-runtime-cu11 11.8.89
nvidia-cuda-runtime-cu12 12.4.127
nvidia-cudnn-cu11        9.1.0.70
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu11        10.9.0.58
nvidia-cufft-cu12        11.2.1.3
nvidia-curand-cu11       10.3.0.86
nvidia-curand-cu12       10.3.5.147
nvidia-cusolver-cu11     11.4.1.48
nvidia-cusolver-cu12     11.6.1.9
nvidia-cusparse-cu11     11.7.5.86
nvidia-cusparse-cu12     12.3.1.170
nvidia-nccl-cu11         2.21.5
nvidia-nccl-cu12         2.21.5
nvidia-nvjitlink-cu12    12.4.127
nvidia-nvtx-cu11         11.8.86
nvidia-nvtx-cu12         12.4.127
onnx                     1.17.0
onnxruntime              1.20.1
packaging                24.2
pillow                   10.2.0
pip                      23.3.1
piper-phonemize          1.1.0
piper_train              1.0.0        /home/turner/PiperTTS/src/python
platformdirs             4.3.6
pooch                    1.8.2
propcache                0.2.0
protobuf                 5.28.3
pycparser                2.22
pyDeprecate              0.3.2
pytorch-lightning        1.9.5
PyYAML                   6.0.2
requests                 2.32.3
scikit-learn             1.5.2
scipy                    1.14.1
setuptools               75.6.0
six                      1.16.0
soundfile                0.12.1
soxr                     0.5.0.post1
sympy                    1.13.1
tensorboard              2.18.0
tensorboard-data-server  0.7.2
threadpoolctl            3.5.0
torch                    2.0.0+cu118
torchaudio               2.0.0+cu118
torchmetrics             0.11.4
tqdm                     4.67.0
triton                   2.0.0
typing_extensions        4.12.2
urllib3                  2.2.3
Werkzeug                 3.1.3
wheel                    0.45.1
yarl                     1.18.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~


14. you can run this script to confirm your cuda version within your VENV
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
import torch
import os

def print_cuda_details():
    if torch.cuda.is_available():
        # GPU device details
        device_id = torch.cuda.current_device()
        device_name = torch.cuda.get_device_name(device_id)
        total_memory = torch.cuda.get_device_properties(device_id).total_memory / (1024**3)  # GB
        memory_allocated = torch.cuda.memory_allocated(device_id) / (1024**3)  # GB
        memory_cached = torch.cuda.memory_reserved(device_id) / (1024**3)  # GB
        multiprocessors = torch.cuda.get_device_properties(device_id).multi_processor_count
        
        # CUDA and PyTorch versions
        cuda_version = torch.version.cuda
        pytorch_version = torch.__version__

        # Device properties
        device_properties = torch.cuda.get_device_properties(device_id)
        compute_capability = device_properties.major, device_properties.minor
        
        # Display the details with improved formatting
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
        print(f"")
        print("Checking CUDA information within the python environment...")
        print(f"")

        print("###########################################################")
        print(f"CUDA Available:         TRUE")
        print(f"Using GPU:              {device_name}")
        print("###########################################################")
        print(f"")

        print(f"CUDA Version:           {cuda_version}")
        print(f"PyTorch Version:        {pytorch_version}")
        print(f"Device ID:              {device_id}")
        print(f"Total Memory:           {total_memory:.2f} GB")
        print(f"Memory Allocated:       {memory_allocated:.2f} GB")
        print(f"Memory Cached:          {memory_cached:.2f} GB")
        print(f"Multiprocessors:        {multiprocessors}")
        print(f"Compute Capability:     {compute_capability[0]}.{compute_capability[1]}")
        print(f"")

        # Device Name
        print(f"CUDA Device Name:       {torch.cuda.get_device_name(device_id)}")
        print(f"")

        # CUDA Path (from environment variables)
        cuda_path = os.getenv('CUDA_PATH', 'Not set')
        print(f"CUDA Path:              {cuda_path}")
        
        # CUDA Toolkit Directory (check via environment variable if available)
        cuda_toolkit_dir = os.getenv('CUDA_HOME', 'Not set')
        print(f"CUDA Toolkit Dir:       {cuda_toolkit_dir}")
        
        # Environment Variables
        print(f"\n")
        print("=================================================")
        print("--- Below are OS SYSTEM Environment Variables ---")
        print(" not just whats available within the python VENV")
        print("=================================================")
        print(f"")

        print(f"{'CUDA_HOME:':<25} {os.environ.get('CUDA_HOME', 'Not set')}")
        print(f"")

        print(f"{'CUDA_VISIBLE_DEVICES:':<25} {os.environ.get('CUDA_VISIBLE_DEVICES', 'Not set')}")
        print(f"")

        print(f"{'LD_LIBRARY_PATH:':<25} {os.environ.get('LD_LIBRARY_PATH', 'Not set')}")
        print(f"")

        print(f"{'PATH:':<25} {os.environ.get('PATH', 'Not set')}")
        print(f"")

        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
    else:
        print("##############################################################")
        print("CUDA is NOT available! You may need to reinstall Pytorch/Torch")
        print("Falling back to using CPU.")
        print("##############################################################")
        print(f"")

        # Provide details for CPU-based configurations
        print(f"PyTorch Version:        {torch.__version__}")
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")

# Call the function to print details
print_cuda_details()
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

which outputs this
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Checking CUDA information within the python environment...

CUDA Available:         TRUE
Using GPU:              NVIDIA GeForce RTX 4060 Laptop GPU

CUDA Version:           11.8
PyTorch Version:        2.0.0+cu118
Device ID:              0
Total Memory:           8.00 GB
Memory Allocated:       0.00 GB
Memory Cached:          0.00 GB
Multiprocessors:        24
Compute Capability:     8.9
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

15. you should still be in the "/traning/piper/src/python" folder, 
if not, change to it:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cd ~/traning/piper/src/python 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

16. run the build_monotonic_align.sh script 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./build_monotonic_align.sh
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by thelabcat** [comment]
`./build_monotonic_align.sh` fails:
```
/home/wilbur/bin/pipertrain/piper/src/python/piper_train/vits/monotonic_align/core.c:18:10: fatal error: Python.h: No such file or directory
   18 | #include "Python.h"
      |          ^~~~~~~~~~
compilation terminated.
error: command '/usr/bin/gcc' failed with exit code 1
```
